### DISCUSSION TITLE
Docs, generics etc

### DISCUSSION THREAD

**Discussion by phh** [larastan/larastan#2000] [Q&A] [answered]
Hi there!

First of all, the work done on generics to Builders, Factories and so on it just *chefs kiss* 👨‍🍳. Really awesome!

Now I'm in the process of updating dependencies from the last couple of months and I've got alot of generic errors - which is awesome since I can now get better IDE completion!

However, I tend to look up previous PRs, comments and searching my way through the internet to find examples of what to specify and how I can do it.

Seems like the [Laravel docs](https://laravel.com/docs/11.x/eloquent-factories) does not provide any examples of generics and I cannot really find anything in Larastan.

Is there a good reference for the generics supported by Laravel - either by framework or with PHPStan?

Thanks 🙏

**Comment by szepeviktor** [comment]
Hello @phh 👋🏻 

Running Larastan on a higher `--level=` will error out if you do not specify [generics](https://www.google.com/search?q=phpstan+generics) in your code.

**Reply by phh** [reply]
Yeah it does. However, having examples like [here](https://github.com/laravel/framework/pull/51851) is super nice.

Using a `HasBuilder` with generics for my models using a custom builder is not really something I would have thought of without reading the release notes (which would be pretty hidden for newcomers in a year or two):

```
/** @use HasBuilder<CommonBuilder<static>> */
use HasBuilder;
```

**Reply by calebdw** [reply]
I can try to add better examples to the docs

**Reply by phh** [reply]
Thanks!

Looking at https://github.com/larastan/larastan/pull/1990 it seems like the examples are really out there. Would just be nice to have a page for those :)

**Reply by phh** [reply]
Hi @calebdw . As of today I needed to find some examples on this, however again I resorted to look into old projects.

I guess I could also help out myself. What would you think of adding an `example.md` in the docs folder, containing examples I've linked to?

**Reply by calebdw** [reply]
@phh, the docs for the Relations has been added [in that PR](https://github.com/larastan/larastan/pull/1990/files#diff-aff15ced1fe37506f41f2e1d93a4d8312529259f7803f8634ffb5eac4de47823)

I'll open a PR to add the builder/factory/collection docs

**Reply by calebdw** [reply, marked as answer]
See https://github.com/larastan/larastan/pull/2036
